File: python_vs_julia/count_triangles.py

import networkx

#### function for counting undirected cycles
def generate_triangles(nodes):
    visited_ids = set() # mark visited node
    for node_a_id in nodes:
        temp_visited = set() # to get undirected triangles
        for node_b_id in nodes[node_a_id]:
            if node_b_id == node_a_id:
                # to prevent self-loops, if your graph allows self-loops then you don't need this condition
                raise ValueError 
            if node_b_id in visited_ids:
                continue
            for node_c_id in nodes[node_b_id]:
                if node_c_id in visited_ids:
                    continue    
                if node_c_id in temp_visited:
                    continue
                if node_a_id in nodes[node_c_id]:
                    yield(node_a_id, node_b_id, node_c_id)
                    #visited_ids.add((node_a_id, node_b_id, node_c_id))
                else:
                    continue
            temp_visited.add(node_b_id)
        visited_ids.add(node_a_id)


node_number   = 100
initial_nodes = 2
G = networkx.barabasi_albert_graph(node_number, initial_nodes, seed=1) 

adj_list = G.adjacency_list() 
n_nodes  = len(adj_list)
adj_dict = {i:adjacent_i for i,adjacent_i in zip(range(n_nodes), adj_list)}

cycles = generate_triangles(adj_dict)
cycles = list(cycles)
print(cycles)

File: python_vs_julia/count_triangles.jl
### Count triangles

function generate_triangles(nodes)
    result = []
    visited_ids = Set{Int}()
    for (node_a_id, nodes_node_a_id) in nodes
        temp_visited = Set()
        for node_b_id in nodes_node_a_id
            if node_b_id == node_a_id
                println("ERROR")
                break
            end
            if node_b_id in visited_ids
                continue
            end
            for node_c_id in nodes[node_b_id]
                if node_c_id in visited_ids
                    continue
                end
                if node_c_id in temp_visited
                    continue
                end
                if node_a_id in nodes[node_c_id]
                    push!(result, [node_a_id, node_b_id, node_c_id])
                else
                    continue
                end
            end        
            push!(temp_visited, node_b_id)
        end
        push!(visited_ids, node_a_id)
    end
    return result
end


function keys_to_ints(dict)
    aux = Dict{Int,Array{Int}}()
    for (k,v) in dict
       aux[parse(k)] = v
    end
    return aux
end

using JSON

aux = readlines("graph.json")
adj_dict = JSON.parse(aux[1])
adj_dict = keys_to_ints(adj_dict)

t0 = time()
println("Start Computing")
triangles = generate_triangles(adj_dict)
println("Number of triangles:", length(triangles))
println("Total time", abs(time()-t0)) 


